Align collection jobs, aggregate shares w RFC 9205 (#675)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

} CollectionJobReq;
~~~

* `query`, the Collector's query. The content of this field depends on the

>   indicated batch mode ({{batch-modes}}).

Reviewer: Please double check that {{batch-modes}} has a reference to {{extending-this-doc}}.

Author: It does, in its very first paragraph.

* `agg_param`, an aggregation parameter for the VDAF being executed. This is

>   the same value as in `AggregationJobInitReq` (see {{leader-init}}).

Author: Do you mean that the line "This is the same value as in `AggregationJobInitReq` (see {{leader-init}})." isn't relevant anymore or that my statement from last week isn't relevant anymore?

collection job. This is because, in general, the aggregation parameter is not
known until this point. In certain situations it is possible to predict the
aggregation parameter in advance. For example, for Prio3 the only valid
aggregation parameter is the empty string.

Upon receipt of a `CollectionJobReq`, the Leader begins by checking that it
recognizes the task ID. If not, it MUST abort with error `unrecognizedTask`.

Next, the Leader checks that the indicated batch mode matches the task's batch
mode. If not, then the Leader MUST abort with error `invalidMessage`.

Next, the Leader checks that the aggregation parameter is valid as described in
{{agg-param-validation}}. If not, then the Leader MUST abort with error
`invalidAggregationParameter`.

If aggregation is performed eagerly, the Leader MUST validate that the
aggregation parameter received in the `CollectionJobReq` matches the aggregation
parameter used in aggregations.

Changing a collection job's parameters is illegal, so if there are further PUT
requests to the collection job with a different `CollectionJobReq`, the Leader

> MUST abort with a client error.

Reviewer: What is a "client error"?

Author: HTTP status in the 400 range. I think it's clear enough from the context that we're discussing HTTP requests, and we lay out what it means to "abort with..." in [the section on errors](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-3.4).

Reviewer: I guess what I'm asking for is to be prescriptive about what error to use from that table.

Author: OK, I think that's fair. I also need to do a pass here to make sure we distinguish clearly between those conditions that get checked during the handling of the PUT request regardless of whether the job will be run asynchronously (so stuff like "is the message well-formed" that's cheap) vs. checks that you'd want to defer (so stuff like "is the right agg param used" that requires fetching a bunch of data from storage).

collection job. This is because, in general, the aggregation parameter is not
known until this point. In certain situations it is possible to predict the
aggregation parameter in advance. For example, for Prio3 the only valid
aggregation parameter is the empty string.

The Leader MAY defer handling the collection request. In this case, it indicates
that the collection job is not yet ready by immediately sending an empty
response body. The Leader SHOULD include a Retry-After header field to suggest a
polling interval to the Collector. The Collector then polls the state of the job
by sending GET requests to the collection job. The Collector SHOULD use each
response's Retry-After header field to decide when to try again. The Leader
responds the same way until either the job is ready, from which point it
responds with a `CollectionJobResp` (defined below), or the job fails, from
which point the Leader MUST abort with the error that caused the failure.

The Leader MAY also wait to respond to the Collector's PUT until the collection
job is ready, in which case it responds with the `CollectionJobResp`, or the job
fails, in which case the Leader MUST abort with the error that caused the

> failure.

Author: 0. Sure, your phrasing is more explicit. I made similar changes to corresponding text in aggregate share handling.
1. Not explicitly, but I added a line of text later indicating when the job should be considered ready. And made a corresponding change to the aggregate share text.
2. In the paragraph above, we introduce it as "`CollectionJobResp` (defined below)". I can hoist up the definition of that structure but I think it's clear enough as is.

The Leader begins handling a `CollectionJobReq` by checking that it recognizes
the task ID. If not, it MUST fail the collection job with error
`unrecognizedTask`.

Next, the Leader checks that the indicated batch mode matches the task's batch
mode. If not, it MUST fail the job with error `invalidMessage`.

Next, the Leader checks that the aggregation parameter is valid as described in
{{agg-param-validation}}. If not, it MUST fail the job with error
`invalidAggregationParameter`.

> If aggregation is performed eagerly, then the Leader checks that the aggregation

Author: It's discussed in [{#aggregate-flow}](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/6d3be110d1422385944664d21ed5a186e5b4ff28/draft-ietf-ppm-dap.md?plain=1#L1680), several sections before this text. If we want to make the discussion of eager aggregation more robust, I think we should do that in a distinct PR.

> If aggregation is performed eagerly, then the Leader checks that the aggregation

Reviewer: What about just sticking in a backwards reference to that section? Though I can see an argument for making this a section, say in operational considerations? (And yeah if we do this, let's do it in a separate PR.)

> If aggregation is performed eagerly, then the Leader checks that the aggregation

Author: We'll come back to this in #682.

collection job. This is because, in general, the aggregation parameter is not
known until this point. In certain situations it is possible to predict the
aggregation parameter in advance. For example, for Prio3 the only valid
aggregation parameter is the empty string.

Upon receipt of a `CollectionJobReq`, the Leader begins by checking that it
recognizes the task ID. If not, it MUST abort with error `unrecognizedTask`.

Next, the Leader checks that the indicated batch mode matches the task's batch
mode. If not, then the Leader MUST abort with error `invalidMessage`.

Next, the Leader checks that the aggregation parameter is valid as described in
{{agg-param-validation}}. If not, then the Leader MUST abort with error
`invalidAggregationParameter`.

> If aggregation is performed eagerly, the Leader MUST validate that the
> aggregation parameter received in the `CollectionJobReq` matches the aggregation
> parameter used in aggregations.

Reviewer: ```suggestion
If aggregation is performed eagerly, the Leader MUST validate that the
aggregation parameter received in the `CollectionJobReq` matches the aggregation
parameter used in each aggregation job pertaining to the batch.
```

Also, what causes this check to fail? Would this be a internal error on the part of the Leader?

Author: It's hard to say. If aggregation was performed eagerly, then either:

- the aggregation parameter for the VDAF is the unit type (`()` in Rust), like for Prio3
- the VDAF does have a non-unit aggregation parameter, but for some reason (special knowledge at the application layer above DAP) its value is known to the Leader such that it could perform eager aggregation.

In the first case, it's not possible for the aggregation parameter values to disagree because there's no aggregation parameter values.

In the second case, I guess it would make sense for the Leader to check that the aggregation parameter it gets from the Collector matches what it guessed when it ran the aggregation jobs.

That's me reaching for a reason to have this check, though.

Reviewer: I think validation makes sense, but let's say waht happens when the check fails. I.e., instead of "MUST validate" write "if validation fails, then MUST abort with error BLAH".

Changing a collection job's parameters is illegal, so further PUT requests to

> the collection job with a different `CollectionJobReq` MUST fail.

Reviewer: Let's replace "MUST fail" with a specific action, like an abort.

Author: Sure. I made a similar change to the equivalent text in the aggregation job init section.

After receiving the `CollectionJobReq`, the Leader begins working with the
Helper to aggregate the reports satisfying the query (or continues this process,

> depending on the VDAF) as described in {{aggregate-flow}}.

Reviewer: ```suggestion
depending on whether the Leader is aggregating eagerly) as described in {{aggregate-flow}}.
```

The Leader first checks whether it can construct a valid batch for the
collection job by applying the requirements in {{batch-validation}}. If so, then
the Leader obtains the Helper's aggregate share following the aggregate-share
request flow described in {{collect-aggregate}}. If not, it either aborts the

> collection job or tries again later, depending on which requirement in

Reviewer: By "abort the collection job" do you mean abort the PUT request from the Collector?

Author: The former. The Leader's validation of  doesn't necessarily happen synchronously during handling of the PUT from the Collector. So I think we should avoid the verb "abort" here. I'll be more explicit here and reuse the language from later where we describe what to do if getting the aggregate share fails.

{{batch-validation}} was not met.

> If the Leader has a pending aggregation job that overlaps with the batch and
> aggregation parameter for the collection job, the Leader MUST first complete the

Reviewer: Multi-collection is no longer possible, so I actually think it's unnecessary to say this.
```suggestion
If the Leader has a pending aggregation job that overlaps with the batch for the collection job, the Leader MUST first complete the
```

aggregation job before proceeding and requesting an aggregate share from the
Helper. This avoids a race condition between aggregation and collection jobs
that can yield batch mismatch errors.

> A collection job's results are represented by a `Collection`, which is

Reviewer: I think I prefer `CollectionJobResp`, for symmetry with aggregation jobs.

structured as follows:

~~~ tls-presentation
struct {
  PartialBatchSelector part_batch_selector;
  uint64 report_count;
  Interval interval;
  HpkeCiphertext leader_encrypted_agg_share;
  HpkeCiphertext helper_encrypted_agg_share;
} Collection;
~~~

> A `Collection`'s media type is "application/dap-collection". The structure

Reviewer: Likewise here: dap-collection-job-resp might be better than dap-collection

* `helper_encrypted_agg_share`: The Helper's aggregate share, encrypted to the
  Collector (see {{aggregate-share-encrypt}}).

> 

Reviewer: Self-note: got here

The Leader then combines the values inside the batch bucket as follows:

* Aggregate shares are combined via `Vdaf.merge(agg_param, agg_shares)` (see
  {{!VDAF, Section 4.4}}), where `agg_param` is the aggregation parameter
  provided in the `CollectionJobReq`, and `agg_shares` are the (partial)
  aggregate shares in the batch buckets. The result is the Leader aggregate
  share for this collection job.
* Report counts are combined via summing.
* Checksums are combined via bitwise XOR.

A Helper aggregate share is identified by a 16-byte ID:

~~~ tls-presentation
opaque AggregateShareID[16];
~~~

The Helper's aggregate share is an HTTP resource served by the Helper at the URL
`{helper}/tasks/{task-id}/aggregate_shares/{aggregate-share-id}`. To obtain it,
the Leader first chooses an aggregate share ID, which MUST be unique within the
scope of the corresponding DAP task.

> Then the Leader sends a PUT request to the aggregate share with the body:

Author: The aggregate share ID is new, and we need it so that the leader can poll on something for asynchronous agg share handling. The alternative would be to not specify an aggregate share ID and let the helper choose a unique URL for the aggregate share, but we chose long ago to have HTTP clients choose resource identifiers (i.e. leader chooses agg job and agg share IDs, collector chooses collection job IDs).

Author: We discussed this at the last PPM interim. What I proposed was to let the leader choose it and nobody argued against that ([1]). My rationale was that we should be consistent with all the other interactions where HTTP clients always choose identifiers (clients choose report IDs, leaders choose aggregation job IDs, collectors choose collection job IDs).

[1]: https://datatracker.ietf.org/meeting/interim-2025-ppm-01/materials/slides-interim-2025-ppm-01-sessa-tim-geoghegan-aligning-dap-with-rfc-9205-00

} AggregateShareReq;
~~~

The media type of the request is "application/dap-aggregate-share-req". The
structure contains the following parameters:
The media type of `AggregateShareReq` is "application/dap-aggregate-share-req".
The structure contains the following parameters:

* `batch_selector`: The "batch selector", the contents of which depends on the
  indicated batch mode: for time-interval mode, the `config` field contains the
  time interval selected by the Collector ({{time-interval-batch-mode}}); in
  leader-selected mode, the field contains the batch ID selected by the Leader

>   ({{leader-selected-batch-mode}}).

Author: Defer discussion of batch modes to {{batch-modes}}

  Future documents that new batch modes MUST specify the contents of the
  `config` field; see {{extending-this-doc}} for details.

  The indicated batch mode MUST match the task's batch mode. Otherwise, the

>   Helper MUST abort with "invalidMessage".

Author: Put all the validation MUSTs in one place, below

Leaders MUST authenticate their requests to Helpers using a scheme that meets
the requirements in {{request-authentication}}.

The Helper first ensures that it recognizes the task ID. If not, it MUST abort
with error `unrecognizedTask`.

The indicated batch mode MUST match the task's batch mode. If not, the Helper
MUST abort with `invalidMessage`.

The Helper then verifies that the request meets the requirements in
{{batch-validation}}. If not, it aborts with the indicated error.

The aggregation parameter MUST match the aggregation parameter used in
aggregation jobs pertaining to this batch. If not, the Helper aborts with

> `invalidMessage`.

Author: Break each validation MUST into its own paragraph for clarity.

The aggregation parameter MUST match the aggregation parameter used in
aggregation jobs pertaining to this batch. If not, the Helper aborts with
`invalidMessage`.

Next, the Helper retrieves and combines the batch buckets associated with the
request using the same process used by the Leader (as described at the beginning
of this section {{collect-aggregate}}), arriving at its final aggregate share,
report count, and checksum values. If the Helper's computed report count and
checksum values do not match the values provided in the `AggregateShareReq`, it
MUST abort with an error of type "batchMismatch".
request using the same process used by the Leader (described at the beginning of
this section), arriving at its aggregate share, report count, and checksum
values. If the Helper's computed report count and checksum values do not match
the values provided in the `AggregateShareReq`, it MUST abort with an error of
type `batchMismatch`.

The Helper then encrypts `agg_share` under the Collector's HPKE public key as
described in {{aggregate-share-encrypt}}, yielding `encrypted_agg_share`.
Encryption prevents the Leader from learning the actual result, as it only has

> its own aggregate share and cannot compute the Helper's.

Author: I deleted this for brevity but could be convinced to put it back.

Author: That's a fair point. I put this back.

The Helper responds to the Leader with HTTP status code 200 OK and a body

> consisting of an `AggregateShare`, with media type

Author: Don't prescribe HTTP status

Leaders MUST authenticate their requests to Helpers using a scheme that meets
the requirements in {{request-authentication}}.

The Helper MAY defer handling the aggregate share request. In this case, it
indicates that the aggregate share is not yet ready by immediately sending an
empty response body. The Helper SHOULD include a Retry-After header field to
suggest a polling interval to the Leader. The Leader then polls the state of the
job by sending GET requests to the aggregate share. The Leader SHOULD use each

> response's Retry-After header field to decide when to try again. The Helper

Author: Seems reasonable, but there's probably other standard HTTP headers we refer to that want the same treatment. How about if we punt to a later PR? I promise I'll do it, it's just easier to extract that from a tricky change like this.

https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/681

computed above and `encrypted_aggregate_share.ciphertext` is the ciphertext
`encrypted_agg_share` computed above.

The Helper's handling of this request MUST be idempotent. That is, if multiple
identical, valid `AggregateShareReq`s are received, they should all yield the

> same response.

Author: The request is now a PUT, which implies idempotence in HTTP, so we don't need this remark

a polling interval to the Leader. The Leader then polls the state of the job by
sending GET requests to the aggregate share. The Helper responds the same way
until the job is ready, from which point it responds with the encoded
`AggregateShare`. The Leader SHOULD use each response's Retry-After header field
to decide when to try again.

The Helper MAY also block responding to the Leader's PUT until the aggregate
share is ready. In this case, it responds with the `AggregateShare`.

After receiving the Helper's response, the Leader includes the HpkeCiphertext in

> its response to the Collector (see {{collect-finalization}}).

Author: I found it awkward to imply that the Leader uses the Helper's agg share for anything, as well as the implication that the leader finalizes the collection job. Seems like that happens when the collector unshards the agg shares.

    agg_share)
~~~

* `pk` is the Collector's HPKE public key
* `server_role` is the Role of the encrypting server (`0x02` for the Leader and
  `0x03` for a Helper)
* `0x00` represents the Role of the recipient (always the Collector)

> * `agg_share_aad` is an `AggregateShareAad` (defined below).

Author: explicitly enumerate the variables for clarity.

When the Leader receives a `Query` in the request from the Collector during
collection initialization ({{collect-init}}), it must first check that the
batch determined by the query can be collected. The Helper performs the same
check when it receives a `BatchSelector` in the request from the Leader for its
aggregate share ({{collect-aggregate}}).

First, the Aggregator checks if the request (the `CollectionJobReq` for the
Leader and the `AggregateShareReq` for the Helper) identifies a valid set of
batch buckets ({{batch-buckets}}). If not, it MUST abort the request with
`batchInvalid`.

Next, the Aggregator checks that the number of reports in the batch is equal to
or greater than the task's minimum batch size. If not, then the Helper MUST
abort with error `invalidBatchSize`. The Leader SHOULD wait for more reports to

> be validated and try the collection job again later.

Author: I think that explaining the comparison to minimum batch size is perfectly clear and more concise than the IMO awkward definition of `X`.

Next, the Aggregator checks if any of the batch buckets identified by the

> request have already been collected. If so, it MUST abort with "batchOverlap".

Author: *Batches* (which consist of one or more batch buckets) can overlap but I'm pretty sure *batch buckets* can't.

## Time Interval {#time-interval-batch-mode}

The time-interval batch mode is designed to support applications in which
reports are grouped by an interval of time. The Collector specifies a "batch
interval" into which report timestamps must fall.

The Collector can issue queries whose batch intervals are continuous,
monotonically increasing, and have the same duration. For example:

~~~ tls-presentation
[
  struct {
    start = 1659544000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659545000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659546000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659547000,
    duration = 1000,
  } Interval
]

> ~~~

Author: This is longer than the existing example of batch intervals, but uses a syntax we already introduced. I can go either way with it.

1. Batch buckets ({{batch-buckets}}): how reports are assigned to batch
   buckets; how each bucket is identified; and how batch buckets are mapped to
   batches

## Time Interval {#time-interval-batch-mode}

The time-interval batch mode is designed to support applications in which
reports are grouped by an interval of time. The Collector specifies a "batch
interval" into which report timestamps must fall.

The Collector can issue queries whose batch intervals are continuous,

> monotonically increasing, and have the same duration. For example:

Reviewer: ```suggestion
monotonically increasing, and have the same duration. For example, the following sequence of batch intervals satisfies these conditions
```

~~~ tls-presentation
[
  struct {
    start = 1659544000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659545000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659546000,
    duration = 1000,
  } Interval,
  struct {
    start = 1659547000,
    duration = 1000,

>   } Interval

Reviewer: unless this would be inconsistent
```suggestion
  } Interval,
```

### Batch Buckets {#time-interval-batch-buckets}

> Each batch bucket is identified by an `Interval` whose duration is equal to the

Author: Use the `Interval` type so that we implicitly get the definition in terms of `time_precision` as well as the comparison and inclusion semantics.

task's `time_precision`. The identifier associated with a given report is the
unique such interval containing the timestamp of the report. For example, if
the task's `time_precision` is 1000 seconds and the report's timestamp is
1729629081, the relevant batch bucket identifier is

~~~ tls-presentation
struct {
  start = 1729629000,
  duration = 1000,
} Interval
~~~

The `Query` received by the Leader or `BatchSelector` received by the Helper
determines a valid set of batch bucket identifiers if
`batch_interval.duration >= time_precision`.

A batch consists of a sequence of contiguous batch buckets. That is, the set of
batch bucket identifiers for the batch interval is

~~~ tls-presentation
[
  struct {
    start = batch_interval.start,
    duration = time_precision,
  } Interval,
  struct {
    start = batch_interval.start + time_precision,
    duration = time_precision,
  } Interval,
  ...
  struct {
    start = batch_interval.start + batch_interval.duration - time_precision,
    duration = time_precision,
  } Interval
]

> ~~~

Author: As above, we use an existing syntax to represent these intervals, at the cost of more vertical space.

chosen by the Leader, which MUST be unique in the scope of the task.

~~~ tls-presentation
opaque BatchID[32];
~~~

The Collector will not know the set of batch IDs available for collection. To
get the aggregate of a batch, the Collector issues a query, which does not
include any information specifying a particular batch (see {{batch-modes}}). The
Leader selects a recent batch to aggregate. The Leader MUST select a batch that
has not yet been associated with a collection job.
get the aggregate of a batch, the Collector issues a query for the next
available batch. The Leader selects a recent batch to aggregate which MUST not
yet have been associated with a collection job.

The Aggregators can output batches of any size that is larger than or equal to
`min_batch_size`. The target batch size, if any, is implementation-specific, and
may be equal to or greater than the minimum batch size. Deciding how soon
batches should be output is also implementation-specific. Exactly sizing batches
may be challenging for Leader deployments in which multiple, independent nodes
running the aggregate interaction (see {{aggregate-flow}}) need to be

> coordinated.

Author: Seems unhelpful to me to say that this part of the protocol is hard without saying what to do about it.

Reviewer: we're saying that coordination in a massively distributed leader may make it impossible to hit the target batch size exactly.

Author: Fair enough. I put the original paragraph back.

### Batch Buckets {#time-interval-batch-buckets}

Each batch bucket is identified by an `Interval` whose duration is equal to the
task's `time_precision`. The identifier associated with a given report is the
unique such interval containing the timestamp of the report. For example, if
the task's `time_precision` is 1000 seconds and the report's timestamp is
1729629081, the relevant batch bucket identifier is

~~~ tls-presentation
struct {
  start = 1729629000,
  duration = 1000,
} Interval
~~~

The `Query` received by the Leader or `BatchSelector` received by the Helper
determines a valid set of batch bucket identifiers if

> `batch_interval.duration >= time_precision`.

Reviewer: Is `batch_interval` defined?

Author: No. I rewrote this in prose without `batch_interval` and I think it's clear.

A batch consists of a sequence of contiguous batch buckets. That is, the set of
batch bucket identifiers for the batch interval is

~~~ tls-presentation
[
  struct {
    start = batch_interval.start,
    duration = time_precision,
  } Interval,
  struct {
    start = batch_interval.start + time_precision,
    duration = time_precision,
  } Interval,
  ...
  struct {
    start = batch_interval.start + batch_interval.duration - time_precision,
    duration = time_precision,

>   } Interval

Reviewer: ```suggestion
  } Interval,
```

]
~~~

## Leader-selected Batch Mode {#leader-selected-batch-mode}

The leader-selected batch mode is used when it is acceptable for the Leader to
arbitrarily batch reports. Each batch is identified by an opaque "batch ID"
chosen by the Leader, which MUST be unique in the scope of the task.

~~~ tls-presentation
opaque BatchID[32];
~~~

The Collector will not know the set of batch IDs available for collection. To
get the aggregate of a batch, the Collector issues a query for the next

> available batch. The Leader selects a recent batch to aggregate which MUST not

Reviewer: ```suggestion
available batch. The Leader selects a recent batch to aggregate which MUST NOT
```
